### PR TITLE
Handle exceptions thrown by user callback

### PR DESCRIPTION
**Pull request by ronag**
Currently it is assumed that any callback sent in by the user NEVER throws. If it does throw then the deepstream state is pretty much undefined.

This is WIP.

### CODE REVIEW

**Comment by ronag** [comment]
For now I've just identified potential problem areas.

@yasserf: How should this be handled, especially in the asynchronous case? `client.emit( 'error', ...)`?

File: src/record/record.js [hunk 1]
```diff
@@ -453,9 +459,11 @@ Record.prototype._onRead = function( message ) {
 Record.prototype._setReady = function() {
 	this.isReady = true;
 	for( var i = 0; i < this._queuedMethodCalls.length; i++ ) {
+		// TODO: What if callback throws?
 		this[ this._queuedMethodCalls[ i ].method ].apply( this, this._queuedMethodCalls[ i ].args );
```
Comment (ronag) [inline src/record/record.js]: This will never throw if we properly validate all the arguments in `set` before enqueueing it.

Comment (ronag) [inline src/record/record.js]: With #221 this wouldn't be a problem.

**Comment by ronag** [comment]
```javascript
tryCatchEmit(client, callback) {
  try {
    callback();
  } catch ( err ) {
    client.emit( 'error', /* how should this look? */);
  }
}
```

**Comment by ronag** [comment]
What I'm mostly unsure about is how to handle the `emit` callbacks. Not sure how NodeJS resolves this issue.

**Comment by ronag** [comment]
It seems the practice in Node is to try/catch the emit and let the user shoot themselves if they have a throwing listeners, in terms of having all listeners called.

**Comment by yasserf** [comment]
do you mean we put the try catch around the for loop then, and if one of them fail it just doesn't notify the rest for that event?

**Comment by ronag** [comment]
@yasserf: That seems to be the practice for internal NodeJS code. However, we should still do `client.emit('error', 'uncaught exception')` or something.

**Comment by ronag** [comment]
If we want to catch for each callback we would have to modify `component-emitter`.

**Comment by yasserf** [comment]
yeah that would be way too much of a performance hit, I like the idea of wrapping the loop, but that in itself is also expensive if ( like in your case ) you have records being requested and discarded constantly

**Comment by ronag** [comment]
>but that in itself is also expensive

I would prefer reliable state over performance in this case.

**Comment by ronag** [comment]
@yasserf: Do you know of any interesting optimisation from V8 in regards to try/catch the last year?

**Comment by yasserf** [comment]
No, the last time I checked this still makes the method unoptimized

**Comment by AlexBHarley** [comment]
Closing due to PR age

**Pull request closed without merging**